plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/contributors/FirCallableCompletionContributor.kt

```diff
@@ -101,7 +100,16 @@ internal open class FirCallableCompletionContributor(
         val availableNonExtensions = collectNonExtensions(implicitScopes, visibilityChecker, scopeNameFilter) { filter(it) }
         val extensionsWhichCanBeCalled = collectSuitableExtensions(implicitScopes, extensionChecker, visibilityChecker)
 
-        availableNonExtensions.forEach { addCallableSymbolToCompletion(context, it, getOptions(it)) }
+        availableNonExtensions
+            // skip shadowed variable or properties
```

Comment:
Hm. Is it correct?
We should see most specific candidate here instead of the first from the scope.

Eg:
```
val Number.a: Int get() = 10
val Int.a: Int get() = 10

fun x() {
    with(1) {
        a<caret> // here we should resolve to Int.a as more specific
    }
}
```
Here we should use tower resolver logic from compiler somehow.

Reply:
This is doing filtering only on non-extension variable-like members. It would includes both local variables and proper members with current implicit receiver. So with or without receiver, there is always at most one symbol that resolution should go to. There won't be any need to pick a "better" one.

I agree that we should delegate this to the tower resolver. But from my observation the current implementation seems to just do the right thing. Consider
```
class Foo {
    val length = ""
    fun String.foo() {
        val length = true
        len<caret>
    }
}
```
Both the current logic and tower resolver picks candidate from inner most and fallback to the outer one if inner one do not exist. So maybe it's sufficient to keep this as it is?

Reply:
Actually after thinking about it more, it seems the filtering done here should only be applied to unqualified local variables. Ideally completion should offer all other variable-like members, whether they are extension members or normal members. Then, upon user's selection, we should add label and explicit receivers to qualify the access in case it's ambiguous. For example
```
val Int.foo : ...
val Number.foo : ...
val String.foo : ...
fun String.test() {
  with(1) {
    val foo = ...
    fo<caret>
  }
}
```
Here, we should offer all the `foo` and completion should end up with
local `foo` -> `foo`
`Int.foo` -> `this.foo`
`Number.foo` -> `(this as Number).foo`
`String.foo` -> `this@test.foo`

Obviously, to do the above correctly, we will need to rely on tower resolver. One potential way to implement this that I can think of is to enumerate all available implicit receivers and resolve `foo` on them. Then for any members available we figure out the "minimal" label/typecast needed in order to make resolution go to that symbol.

But I am guessing this implementation would be quite slow. And the benefit may not worth the effort since user should really not use the same name for many things. Maybe we can add a TODO here?

Reply:
Yeap, let's leave TODO for now. Looks like this require separate design